### PR TITLE
ssl_options() option inside tls()

### PR DESCRIPTION
**Pull request by pasztor**
allowed values: no-sslv2, no-sslv3, no-tlsv1, no-tlsv11, no-tlsv12, none
I think, they are self-explanatory.

Signed-off-by: PÁSZTOR György <gyorgy.pasztor@balabit.com>

### CODE REVIEW

**Comment by deirf** [comment]
:+1:

File: lib/tlscontext.h [hunk 1]
```diff
@@ -44,6 +44,16 @@ typedef enum
...
+typedef enum
+{
+  TSO_NONE,
+  TSO_NOSSLv2=0x0001,
+  TSO_NOSSLv3=0x0002,
+  TSO_NOTLSv1=0x0004,
+  TSO_NOTLSv11=0x0008,
+  TSO_NOTLSv12=0x0010,
+} TLSSslOptions;
+
```
Comment (juhaszviktor) [inline lib/tlscontext.h]: You use TlsSslOption as a flag, but the type of it is enumeration.
Enumerations are usually used for unique values, not for combinations of these. 
I mean, that you should "typedef unsigned int TlsSslOptions" and declare the TSO_* as define (e.g.: #define TSO_NOSSLv2 1)

File: lib/tlscontext.c [hunk 2]
```diff
@@ -408,6 +428,33 @@ tls_lookup_verify_mode(const gchar *mode_str)
...
+        ret|=TSO_NOTLSv11;
+      else if (strcasecmp(l->data,"no-tlsv12") == 0 || strcasecmp(l->data,"no_tlsv12") == 0)
+        ret|=TSO_NOTLSv12;
+#endif
+      else
+        msg_error("Unknown ssl-option",evt_tag_str("option",l->data),NULL);
+    }
+  msg_debug("ssl-options parsed",evt_tag_printf("parsed value","%d" ,ret),NULL);
+  return ret;
+}
```
Comment (juhaszviktor) [inline lib/tlscontext.c]: Please use the gnu coding style, put space after the comma in the parameter list,
put space after "for" and spaces in the assignment, as you did in the line 396.

File: lib/tlscontext.c [hunk 3]
```diff
@@ -334,7 +335,25 @@ tls_context_setup_session(TLSContext *self)
...
+#ifdef SSL_OP_NO_TLSv1_2
+          if(self->ssl_options & TSO_NOTLSv11)
+            ssl_options |= SSL_OP_NO_TLSv1_1;
+          if(self->ssl_options & TSO_NOTLSv12)
+            ssl_options |= SSL_OP_NO_TLSv1_2;
+#endif
+          SSL_CTX_set_options(self->ssl_ctx, ssl_options);
+        }
+      else
+	msg_debug("empty ssl options",NULL);
```
Comment (juhaszviktor) [inline lib/tlscontext.c]: These lines should be extracted into a different function (e.g.: set_ssl_option or something)

File: lib/tlscontext.h (hunk 1 above)
Comment (bazsi) [inline lib/tlscontext.h]: I would slightly disagree. Enums are available in a debugger while defines
are not.

Comment (juhaszviktor) [inline lib/tlscontext.h]: I can agree your standpoint, but if I look at the whole syslog-ng code, the flags and some kind of options are declared as a define. For example in logwriter.h LW_* and LWO_* are also defines.

Comment (lbudai) [inline lib/tlscontext.h]: And what if we have a `typedef` for `TlsSslOption` and `static const TlsSslOption ` delarations for the values? Or we just change the return type of `tls_lookup_options` to `int`...

I think the problem with `enum` is that when we return from `tls_lookup_options` we say that type of the value is `TLSSslOptions` but it's not: it`s a combination of the enumerated values.

Comment (bazsi) [inline lib/tlscontext.h]: a number of this calls were converted to enums recently. I did start to
move to the enum direction. An example is TLSVerifyMode just above this
declaration. Or msg-format.h, or a number of others. So this is not
consistent. I do know that I was converting a couple of these.

In these cases the call sites do not use the enum type (I don't even see
the use for the typedef itself, that should be abandoned), the the
call-sites should use "int"


-- 
Bazsi

Comment (bazsi) [inline lib/tlscontext.h]: well, const int's is a lot more typing and I would never want to make
actual variables from these. I dunno if the compiler gets rid off those or
not....

I did check, it doesn't optimize those away:

#include <stdio.h>

const int foobar = 1;

int main()
{
  printf("%d", foobar);
}

bazsi@bzorp:~$ gcc -g a.c -S -o a.i

        .globl  foobar
        .section        .rodata
        .align 4
        .type   foobar, @object
        .size   foobar, 4
foobar:
        .long   1

So it is part of .rodata, though the call-site itself uses the literal
value:

        movl    $1, %eax
        movl    %eax, %esi
        movl    $.LC0, %edi
        movl    $0, %eax
        call    printf




-- 
Bazsi

Comment (lbudai) [inline lib/tlscontext.h]: Ok, then we can keep `enum`, eliminate `typedef` and just use an `int` as return type.

Comment (pasztor) [inline lib/tlscontext.h]: I tried to be consequent with the solution what was already there for the verify options. I did pretty much the same.
If my solution is bad, then verify options's is bad too.
If you'd like, I'd gladly correct both of them, but I think that should be a scope of a different PR, and this PR should focus only to implement the ssl-options() setting.
Opinions?

File: lib/tlscontext.c (hunk 3 above)
Comment (pasztor) [inline lib/tlscontext.c]: I thinks, here stands the pretty same, what I just wrote for the typedef part.
I can accept, that this should be do like that, but that should be done in a different PR, which's scope is to reduce the whole tls_context_setup_session()'s size.
As I already wrote there: I would gladly make the custumization for both part, in the way you expect it, but I think the optimal here if we leave this as is, so after the merge it will be still consequent with the very similar verify-options's solution.
And in a separate PR I will do the other alignment's.
Opinions?

File: lib/tlscontext.c [hunk 4]
```diff
@@ -280,6 +280,7 @@ tls_context_setup_session(TLSContext *self)
   SSL *ssl;
   TLSSession *session;
   gint ssl_error;
+  long ssl_options;
```
Comment (gregory094) [inline lib/tlscontext.c]: I know no reason why this option should be `long`.

File: lib/tlscontext.c (hunk 3 above)
Comment (gregory094) [inline lib/tlscontext.c]: As far as I can see we *transform* our `ssl_options` into `openssl_options`. This transformation could be carried out by declaring  a `compile_table` where the entries are `{ssl_option, openssl_option}` pairs. After that a simple iteration through the elements can do the work.

I also support the idea of @lbudai of declaring `TLSSSLOption` and `TLSOpenSSLOption`.

Sketch:
```C

typedef gint TLSSSLOption;
typedef gint TLSOpenSSLOption;

typdef struct 
{
   TLSSSLOption ssl_option;
   TLSOpenSSLOption openssl_option;
} TLSSSLOptionPair;

TLSSSLOptionPair TLSSSLOptionTransformTable[]= 
{
     { TSONOSSLv2, SSL_OP_N_SSLv2 },
     ...
     ...
     ...
};

........

`ssl_transform_options function`:

for(int i = 0; i < sizeof(TLSSSLOptionTransformTable); ++i)
{
    if ( self->ssl_options & TLSSSLOptionTransformTable[i].ssl_option )
        ssl_options |= TLSSLOptionTransformTable[i].openssl_option;
}

```

File: lib/tlscontext.c (hunk 4 above)
Comment (pasztor) [inline lib/tlscontext.c]: I just say only one: Manual page SSL_CTX_set_options(3ssl)
SYNOPSIS
        #include <openssl/ssl.h>

        long SSL_CTX_set_options(SSL_CTX *ctx, long options);

That internal variable is for prepare the value what SSL_CTX_set_options will get.
Is that satisfactory enough? ;-)

File: lib/tlscontext.c (hunk 3 above)
Comment (pasztor) [inline lib/tlscontext.c]: On the long term, it may reduce the size of code, and would gave us the possibility to simply add new options.
Maybe. Maybe we want to gave the chance through our implementation, to not set, but instead SSL_CTX_clear_options regarding the default set.
Then it may almost double the size of your version of code if it's flexible enough at all.
In that case also harden our work, if you want to make two of syslog-ng's ssl-options option to be mutualy exclusive, etc.
So at this point I would not bother this type of solution, since at this point, with this feature set, this is the most compact version. In the future, if there is real need for other options, the it may have a worth to think about that, but I would say, not now, not with this simple feature set.

File: lib/tlscontext.c (hunk 4 above)
Comment (gregory094) [inline lib/tlscontext.c]: Absolutely. :)

**Comment by deirf** [comment]
fixes #457

File: lib/tlscontext.h (hunk 1 above)
Comment (lbudai) [inline lib/tlscontext.h]: hi,

I'm reluctant to merge the patch in this form.
Please, use `int` as return type, as I suggested before (and you can keep `enum`... ).

L.

Comment (pasztor) [inline lib/tlscontext.h]: hi,

as I said: this is consistent with the other parts from the same module in _this_ form.
I also said, that I agree with you about the typedef enum part.
So, I am wondering if
* you do not trust my word, that I will make a code cleaning for this module where I fix the ssl-options and the verify-mode both
* or you want a code where two very-very similar functions (tls_lookup_verify_mode, tls_lookup_options) are defined two different ways in the same code
* or you want a larger patch, where a code cleanup is done at the same branch where a new function/ option is implemented?

Comment (lbudai) [inline lib/tlscontext.h]: hi,

I told you IRL that this is not about consistency: it's about correctness.
(a legacy code will always contain some legacy code ;-) That doesn't mean we should always follow it
and doesn't mean that it cannot be changed... )

I think 2nd option in your list would be fine with an additional note:
If we have a chance for writing a not-legacy code, we should have to do that, IMHO. 

L.

**Comment by lbudai** [comment]
Thanks!!
